### DISCUSSION TITLE
感谢大佬

### DISCUSSION THREAD

**Discussion by robertenjoy** [askfiy/nvim#2] [General]
感谢大佬的项目，看着知乎，然后根据这个项目终于装上了部分。

在装的过程中遇到的难题
1. 依赖问题
部分插件需要依赖其他的软件，而且很大，如node等。
像我这种无法翻墙或下载速度很慢的感觉很难受，
如果有一个部分说明下这些插件有什么用，需要安装哪些依赖，就好了
这样我就可以选择性的不去装那些插件。

以下是我根据作者知乎找到的部分内容
nvim-lint 代码诊断 需要nodejs和npm 让python下载语言服务器pyright
nvim-treesitter 语法高亮 需要build-essential(gcc工具包) tar curl git
neoformat 代码格式化 需要手动下载各个语言的格式化程序  lua：直接通过 npm 安装 npm install -g lua-fmt

nvim-spectre 当前工作区下跨文件修改某段字符 ripgrep
telescope 模糊搜索文件、文字、文档等 fd ripgrep

2. 快捷键配置文档带一些简略的中文注释就好了，如何只看那个文档有点不知道是什么作用

**Comment by askfiy** [comment]
@robertenjoy 

抱歉，回复的有些晚。

第一点，node 和 npm 我认为是很有必要的，它带来了许多东西

- nvim-lsp-installer 通过 npm 可以非常快速的下载 LSP 服务
- neoformat 通过 npm 可以下载很带代码格式化程序，如 prettier、gofmt 等等

相比于手动的去安装它们，node 能够节约非常多的时间来快速的安装这些软件。

第二点，翻墙的问题。你可以试试 clash for windows，一款非常棒的翻墙软件。

如果实在不能翻墙也没关系，packer 包管理可以设置插件下载源，你可以手动设置从 github 镜像源下载这些插件，而不是从 github 原站。

npm 也可以设置淘宝镜像源，下载也很快。

第三点，依赖问题。在 README 文档中，我已经将这个项目目前使用到的依赖都罗列了出来：

```
-- 让 neovim 能够和系统剪切板交互
$ yay -S xsel

-- 模糊查找相关（telescope 和 spectre）
$ yay -S fd
$ yay -S sed
$ yay -S ripgrep

-- 代码格式化（neoformat 插件所需要的）
$ npm install -g lua-fmt
$ npm install -g prettier
$ npm install -g gofmt
$ pip3 install autopep8
$ pip3 install sqlformat

-- Translate（translate 翻译插件所需要的，你也可以不装这个依赖而使用 deepl_free 所提供的接口，但目前国内不能使用 deepl_free 或者 deepl_pro 所提供的接口，所以针对英文不好的朋友该依赖是必要的）
$ yay -S translate-shell

-- markdown preview（这是 nvim-markdown-preview 所需要的 2 个依赖，能够让你快速的预览 markdown 文档）
$ yay -S pandoc
$ npm install -g live-server

-- go 语言和 python 的调试所需要的插件，如果你不用这 2 种语言你可以不装
$ yay -S delve
$ python3 -m pip install debugpy

-- Python 的代码诊断，你也可以不装这个而使用 Pyright，但这个更强大也更专业
$ pip3 install pylint

--如果你不使用 pylint，而使用 Pyright 诊断 python 代码，恰好你又在编写 Django 程序，那么这个插件是必要的
$ pip3 install django-stubs

-- 这是一款在终端里很方便操作 git 的程序，非常好用
$ yay -S lazygit

-- curl 命令 translate 插件和 tabnine 插件是都需要的，而 unzip 只是 tabnine 需要
$ yay -S unzip
$ yay -S curl
```

第四点，快捷键的注释。因为日后我可能会添加新的插件，快捷键可能会发生变更。所以我不想详细的阐述每一个按键是做什么的，后面改动会比较麻烦。

我创建了一个 QQ 群，如果你有啥问题可以直接加进来问：978088231

**Reply by sbccznfy** [reply]
QQ群已经解散了吗

**Reply by askfiy** [reply]
> QQ群已经解散了吗

Yes ~

**Comment by robertenjoy** [comment]
在windows下aux是设备名，导致git失败。如果新建aux目录会提示‘指定的设备名无效’
以下是查询到的内容
此为Windows操作系统的限制，一些特定字符的组合不能用作文件、文件夹名。
如：aux、com1、com2、prn、con、nul 等，这些都是设备名，为系统保留。

**Reply by askfiy** [reply]
涨知识了，或许我该重新换个名字。

**Comment by robertenjoy** [comment]
plugins.lanaguage 是language吗 是否拼错了？

**Reply by askfiy** [reply]
应该是拼写错了，大意了，感谢指正。

**Comment by robertenjoy** [comment]
欸，太客气了。还是大佬你的项目好啊

mrjones2014/nvim-ts-rainbow 好像也依赖nvim-treesitter 也需要加个after

tami5/sqlite.lua  这个项目好像失效了
kyazdani42/nvim-tree.lua 这个也好像失效了 变成了old-nvim-tree.lua

nvim-telescope/telescope-fzf-native.nvim  这个与nvim-telescope/telescope.nvim 联系太紧密了
如果我没下前者仅使用后者会报错，不知道有没有判断是否下载项目成功的api，
当nvim-telescope/telescope-fzf-native.nvim安装成功后才在telescope.lua使用下面的配置 
        extensions = {
            fzf = {
                fuzzy = true,
                override_generic_sorter = true,
                override_file_sorter = true,
                case_mode = "smart_case",
            },
        },

M.telescope.load_extension("fzf")
当然无关紧要，我只是因为windows下没有cmake环境安装不上才发现的。一堆nvim-treesitter的也装不上^_^。

关于phaazon/hop.nvim的快捷键，官方的案例好像提供了替换默认f/F/t/T的方式。可以在当前行中选择查找到的哪一个字符串，还有也有个'/'(:HopPattern)的，不知道你知道不知道。

有没有什么方式设置多个键的快捷键，比如<ctrl+alt+l> 这种的，有ctrl+alt ？

**Reply by robertenjoy** [reply]
下载tzachar/cmp-tabnine失败了 提示说使用install.lsp 该如何使用啊？没有头绪

**Reply by shellRaining** [reply]
> 下载tzachar/cmp-tabnine失败了 提示说使用install.lsp 该如何使用啊？没有头绪

我使用的是 mac，可以进入 ~/.local/share/nvim/site/pack/packer/start/cmp-tabnine 目录，这里面有 install.sh，运行这个脚本就可以了

**Reply by robertenjoy** [reply]
哦 找到了 谢谢。😊
install.ps1文件

**Reply by askfiy** [reply]
@robertenjoy 彩虹括号加了 after 会失效。

Telescope 我认为加上那个扩展可能性能更好，所以没打算将它们解绑。

NvimTree 这个我最近没太关注，后面有空再弄吧。

Sqlite 这个后面可能我会把他干掉，暂时先这样用着吧。

Hop 替换的 fFtT 我知道这个，但我不想过多替换原生的功能，我觉得目前快捷键能到当前行任意字符上就足够使用啦。

Ctrl+Alt 这种组合键是我讨厌的，我必须将左手移开，我太懒了.

**Reply by robertenjoy** [reply]
@askfiy  原来如此。
对于ctrl+alt这种我也不喜欢，但我是idea用户只能妥协，我不想修改默认键位，且也不想记太多的键位。所以我希望将类似的功能改成同一个快捷键。如果你知道该怎么配置的话，能告诉我就好了。

我记得以前有个全局替换的功能，界面类似于Telescope，是哪一个插件来着，可能插件太多我看走眼忽略了。你把知乎文章一删我有点找不着了。

**Reply by robertenjoy** [reply]
哈 我又来了 

# 对于nvim-treesitter.install 

建议默认使用git下载(不然得改成全局代理)   require("nvim-treesitter.install").prefer_git = true
也就是那个配置文件中
requires添加'nvim-treesitter.install'
M.load()添加 M.nvim_treesitter_install.prefer_git = true

不建议ensure_installed=‘all'  下载全部支持的语言，太多了

# 对于使用wsl的用户
wsl剪切板与系统相通，添加如下代码
if vim.fn.has('wsl') then
    vim.g.clipboard = {
        name='myClipboard',
        copy={
            ['+']='clip.exe', -- 系统
            ['*']={'tmux','load-buffer','-'}, -- tmux
        },
        paste = {
            ['+'] = 'powershell.exe -c [Console]::Out.Write($(Get-Clipboard -Raw).tostring().replace("`r", ""))',
            ['*']={'tmux','save-buffer','-'},
        }
    }
end

**Comment by robertenjoy** [comment]
原来替换是`nvim-pack/nvim-spectre`

`nvim-tree`只是改了作者名字

```lua 
nvim-tree/nvim-tree.lua
```

**Reply by askfiy** [reply]
好的，稍后会更新上去

**Comment by robertenjoy** [comment]
@askfiy 
这个我找了半天解决不了，需要你看下
我把`Mason`的所有`dap`全注解掉了，然后使用`<space>db`，然后就报错，我尝试配了一个，也一样。

```packer.nvim: Error running config for nvim-dap: vim/shared.lua:0: after the second argument: expected table, got nil```

你有这样的错误吗？这部分配置我应该没有怎么动过。

**Reply by robertenjoy** [reply]
顺便附上我改动的地方
`Mason` `dap`添加  `'codelldb'`

添加`rust.lua` 文件
```lua
local port = 13000
return {
    adapters = {
        codelldb = {
            type = "server",
            -- port = "${port}",
            port = port,
            executable = {
                -- CHANGE THIS to your path!
                -- command = "/absolute/path/to/codelldb/extension/adapter/codelldb",
                command = "/home/r/.local/share/nvim/mason/packages/codelldb/extension/adapter/codelldb",
                -- args = { "--port", "${port}" },
                args = { "--port", port },

                -- On windows you may have to uncomment this:
                -- detached = false,
            },
        },
    },
    configuration = {
        rust = {
            name = "Launch file",
            type = "codelldb",
            request = "launch",
            program = function()
                return vim.fn.input("Path to executable: ", vim.fn.getcwd() .. "/", "file")
            end,
            cwd = "${workspaceFolder}",
            stopOnEntry = false,
        },
    },
}
```

**Reply by robertenjoy** [reply]
configurations 少了s。/(ㄒoㄒ)/~~
但还是没有运行 我再找找问题

**Comment by robertenjoy** [comment]
居然这么晚了

因为测试我装了两个lsp 一个mason装的
然后发现 mason注解掉那个lsp还是会加载

get_installed_servers 还是会获取到
然后填入默认值配置进行加载。

**Comment by askfiy** [comment]
因为你已经安装了它，解注释是无用的。发自我的 iPhone在 2023年1月12日，00:15，robertenjoy ***@***.***> 写道：﻿
居然这么晚了
因为测试我装了两个lsp 一个mason装的
然后发现 mason注解掉那个lsp还是会加载
get_installed_servers 还是会获取到
然后填入默认值配置进行加载。

**Comment by robertenjoy** [comment]
稍微做下处理，(●ˇ∀ˇ●)
# nvim-lspconfg.lua
> 没有经过测试 应该没问题把
```lua
    for _, server_name in ipairs(M.mason_lspconfig.get_installed_servers()) do
        -- 遍历配置，当没有此lsp配置时不加载
        local mason_conf = require("config/basic/mason")
        local is_exist_conf = false
        for _, v in ipairs(mason_conf.installer_resources.lsp) do
            if server_name == v then
                is_exist_conf = true
                break
            end
        end
       -- 没有continue就算了毕竟lua确实没有  连goto也没有(提示语法错误，可能因为版本或nvim删了) 那只能写丑一点了
        if is_exist_conf == true then
            local require_path = string.format("%s%s", configurations_dir_path, server_name)

            local ok, settings = pcall(require, require_path)

            if not ok then
                settings = {}
            end

            settings.capabilities = aux_lspconfig.get_capabilities()
            settings.handlers = aux_lspconfig.get_headlers(settings)

            settings.on_attach = function(client, bufnr)
                M.nvim_navic.attach(client, bufnr)
                client.server_capabilities.documentFormattingProvider = false
                -- close semantic tokens
                client.server_capabilities.semanticTokensProvider = nil
            end

            M.lspconfig[server_name].setup(settings)
        end
    end
```

# dap配置 c.lua
- 试了两种c系的dap。codelldb好像无法在wsl使用，cpptools好像可以，还有一个装起来麻烦名字忘了
> 就是简单调试hello world的程序时候，打印没有显示出来，不知道为啥，到底是哪里弄错了
```lua
local cpptools_path = require("mason-registry").get_package("cpptools"):get_install_path() .. "/extension"
local cpptools_bin = cpptools_path .. "/debugAdapters/bin/OpenDebugAD7"

return {
    adapters = {
        codelldb = { -- 不使用
            type = "server",
            port = "${port}",
            executable = {
                -- CHANGE THIS to your path!
                -- command = "/absolute/path/to/codelldb/extension/adapter/codelldb",
                command = codelldb_bin,
                args = { "--port", "${port}" },

                -- On windows you may have to uncomment this:
                -- detached = false,
            },
        },
        cppdbg = {
            id = "cppdbg",
            type = "executable",
            -- command = "/absolute/path/to/cpptools/extension/debugAdapters/bin/OpenDebugAD7",
            command = cpptools_bin,

            -- If you're on Windows, use this definition instead:
            -- options = {
            --     detached = false,
            -- },
        },
    },
    configurations = {
        rust = {

            {
                name = "Launch file",
                type = "cppdbg",
                request = "launch",
                program = function() -- 指定debug的可执行二进制文件 需要在项目根目录下开始编辑文件 不然workSpace路径不对，导致getcwd()返回结果不是想要的
                    return vim.fn.input("Path to executable: ", vim.fn.getcwd() .. "/target/debug/"..vim.fn.fnamemodify(vim.fn.getcwd(), ':p:h:t'), "file")
                end,
                cwd = "${workspaceFolder}",
                stopAtEntry = true,
            },
        },
        c = conf, -- 与上配置类似 需要改program项 (这里并没有定义conf变量)
        cpp = conf,
    },
}

```

**Reply by robertenjoy** [reply]
果然有点问题，直觉没错，居然不一一对应
特例的话稍微修改下
```lua
            if server_name == v then
                is_exist_conf = true
                break
            elseif server_name == 'sumneko_lua' and v == 'lua-language-server' then
                is_exist_conf = true
                break
            end

```

**Reply by robertenjoy** [reply]
`mason`会将执行文件放到其bin目录下，并且在开启`nvim`的时候好像加入到了环境变量。所以指定执行文件的配置倒是不用写那么长。
在`windows`下`java`的`lsp`：`jdtls`的执行文件名是`jdtls.cmd`，一定得加上`cmd`后缀(也不知道为啥不能省略)，然后还得指定`data`(缓存目录)和`configuration`(是什么操作系统的配置，不确定是否是这个英文名字)。`jdtls.cmd`依赖`python3.9`如果没有的话还是用`java -jar`的方式，如果是手动下载的`jdtls`可能没有`jdtls.cmd`文件，那就去下载的`jdtls` `bin`目录下找`jdtls.py`也是一样的，我配的时候最新的`jdtls`提示说需要`jdk17`

**Comment by robertenjoy** [comment]
刚看到了提交记录，居然因为aux改了这么多。
我只是将`aux`文件改成了`_aux` 变量名没有动，只是将`require`导入包的地方改成了`_aux`
都行吧 哈哈😄

**Comment by robertenjoy** [comment]
需要一些简单的说明来知道这些插件是用来干嘛的。在配置文件内再提供更多的使用说明就更好了，好久没用了，导致我是一个一个查过来的。哭。虽然确实很麻烦。
![runoob-code](https://user-images.githubusercontent.com/48754716/211998397-6e3dd693-f70a-4c32-a70a-275249536bb7.png)

**Reply by robertenjoy** [reply]
language分类的第二个看名字一直以为与python有关，不过原来是缩进处理的插件
tools分类的carbon原来是将代码发送到carbon网站上可以将其转换为图片
views分类的spectre是替换忘记改了

**Reply by robertenjoy** [reply]
我还真的以为markdowm-preview能在终端查看，原来是通过pandoc将markdown转换成html，然后通过live-server打开这个网页

**Reply by shellRaining** [reply]
> 我还真的以为markdowm-preview能在终端查看，原来是通过pandoc将markdown转换成html，然后通过live-server打开这个网页

其实是有一个叫做 glow 的命令行工具可以用来在终端查看 markdown 的，而且有一个与之对应的 nvim 插件叫做 glow.nvim，不过这个插件的体验不是很好，还需要再观望一下

**Comment by robertenjoy** [comment]
@askfiy 
需要把language目录的名字也修改下

新增一个插件在里面，死活不生效，找了半天原因没找到，差点哭了😭。

**Comment by askfiy** [comment]
Emmmm，忽略了发自我的 iPhone在 2023年1月17日，17:17，robertenjoy ***@***.***> 写道：﻿
@askfiy
需要把language目录的名字也修改下
新增一个插件在里面，死活不生效，找了半天原因没找到，差点哭了😭。
